nsdb-core/src/main/scala/io/radicalbit/nsdb/statement/StatementParser.scala

```diff
@@ -72,115 +70,83 @@ object StatementParser {
       new Sort(new SortField(order.dimension, sortType, order.isInstanceOf[DescOrderOperator]))
     })
 
-    val expParsed = ExpressionParser.parseExpression(statement.condition.map(_.expression), schema.fieldsMap)
-    val fieldList: Either[String, List[Field]] = statement.fields match {
-      case AllFields() => Right(List.empty)
-      case ListFields(list) =>
-        val metricDimensions     = schema.fieldsMap.values.map(_.name).toSeq
-        val projectionDimensions = list.map(_.name).filterNot(_ == "*")
-        val diff                 = projectionDimensions.filterNot(metricDimensions.contains)
-        if (diff.isEmpty)
-          Right(list)
-        else
-          Left(StatementParserErrors.notExistingDimensions(diff))
-    }
+    val expParsed: Either[String, ExpressionParser.ParsedExpression] =
+      ExpressionParser.parseExpression(statement.condition.map(_.expression), schema.fieldsMap)
 
     val distinctValue = statement.distinct
 
     val limitOpt = statement.limit.map(_.value)
 
-    expParsed match {
-      case Right(exp) =>
-        (distinctValue, fieldList, statement.groupBy) match {
-          case (_, Left(errorMessage), _) => Left(errorMessage)
-          // Trying to order by a dimension not in group by clause
-          case (false, Right(Seq(Field(_, Some(_)))), Some(group))
-              if sortOpt.isDefined && !Seq("value", group.field).contains(sortOpt.get.getSort.head.getField) =>
+    expParsed flatMap { exp =>
+      FieldsParser.parseFieldList(statement, schema).flatMap { fieldsList =>
+        //switch on group by
+        (statement.groupBy, fieldsList.list) match {
+          case (Some(group), _)
+              if sortOpt.exists(sort => !Seq("value", "*", group.field).contains(sort.getSort.head.getField)) =>
```

Comment:
is this failing if i send `select * from metric sort by` ?

Reply:
Good point. That `head` is misleading. That `sortOpt` variable is created as follows 

```
val sortOpt = statement.order.map(order => {
      val sortType = schema.fieldsMap.get(order.dimension).map(_.indexType.sortType).getOrElse(SortField.Type.DOC)
      new Sort(new SortField(order.dimension, sortType, order.isInstanceOf[DescOrderOperator]))
    })
```

the `Sort` class is a lucene class that contains an array of `SortField`

Also, a more general consideration: This parser is a semantic one and it's used to translate a parsed SQL query into a Lucene query.
Consequently, the string `select * from metric order by` which is not a valid SQL is checked before by another parse that rejects it.
Anyway, I will refactor that line in a more elegant way that hides the head usage

```
sortOpt.flatMap(_.getSort.headOption).exists(sort => !Seq("value", "*", group.field).contains(sort.getField))
```

Reply:
I wondered about the parser rejection. Ok for the refactor, thank you